zebra-network/src/peer_set/initialize.rs

```diff
@@ -182,47 +188,149 @@ where
+        ?peerset_initial_target_size,
+        "connecting to initial peer set"
+    );
+
+    let initial_meta_addr = initial_peers.into_iter().map(MetaAddr::new_seed_meta_addr);
+
+    // # Correctness
+    //
+    // ## Concurrency
+    //
+    // We spawn each handshake in a separate task. This avoids:
+    // - sequentially waiting on each handshake's timeout, and
+    // - dependencies between the first successful initial peer and other tasks
+    //
+    // ## Denial of Service
+    //
+    // We sleep for `MIN_PEER_CONNECTION_INTERVAL` between handshakes.
+    //
+    // ## Buffer Reservations
     //
-    // Each `CallAll` can hold one `Buffer` or `Batch` reservation for
-    // an indefinite period. We can use `CallAllUnordered` without filling
-    // the underlying `Inbound` buffer, because we immediately drive this
-    // single `CallAll` to completion, and handshakes have a short timeout.
-    let mut handshakes: FuturesUnordered<_> = initial_peers
-        .into_iter()
-        .map(|addr| {
-            outbound_connector
-                .clone()
-                .oneshot(addr)
-                .map_err(move |e| (addr, e))
-        })
-        .collect();
-
-    while let Some(handshake_result) = handshakes.next().await {
-        // this is verbose, but it's better than just hanging with no output
-        if let Err((addr, ref e)) = handshake_result {
-            info!(?addr, ?e, "an initial peer connection failed");
+    // Each `FuturesUnordered` can hold one `Buffer` reservation for an
+    // indefinite period. We can use `FuturesUnordered` without filling the
+    // underlying `Handshake` buffer, because we immediately drive this single
+    // `FuturesUnordered` to completion, and handshakes have a short timeout.
+    let mut handshakes = FuturesUnordered::new();
+    // TODO: replace with *success_count_tx.borrow() in Tokio 1.6
+    let mut success_count = 0;
+    for candidate in initial_meta_addr {
+        // 1. Spawn a task to handshake with a new peer
+        let next_peer_sleep_until = Instant::now() + constants::MIN_PEER_CONNECTION_INTERVAL;
+        let hs_join = tokio::spawn(dial(candidate, outbound_connector.clone()))
+            .map(move |res| match res {
+                Ok(crawler_action) => crawler_action,
+                Err(e) => {
+                    panic!(
+                        "panic during initial handshake with {:?}: {:?} ",
+                        candidate, e
+                    );
+                }
+            })
+            .instrument(Span::current());
+        handshakes.push(Box::pin(hs_join));
+
+        // 2. Check if any peers have finished their handshakes
+        // If we've had enough successes, wait for all the handshakes to finish
+        while let Some(handshake_action) = if success_count < peerset_initial_target_size {
+            handshakes.next().now_or_never().flatten()
+        } else {
+            handshakes.next().await
+        } {
```

Comment:
This code isn't pretty, but it's hard to refactor, because it has a lot of futures and channels.